Subject: [PATCH] Add windows conan compatibility

* Disable whole program optimization to be portable across different MSVC
  versions. See conan-io/conan-center-index#4811 and
  conan-io/conan-center-index#4094.
---
 win/rules.vc     | 23 +++++++----------------
 2 files changed, 11 insertions(+), 20 deletions(-)

diff --git a/win/rules.vc b/win/rules.vc
--- a/win/rules.vc
+++ b/win/rules.vc
@@ -245,7 +245,7 @@
 # BEGIN Case 2(a) - Building Tcl itself
 
 # Only need to define _TCL_H
-_TCL_H = ..\generic\tcl.h
+_TCL_H = $(_TCLDIR)\include\tcl.h
 
 # END Case 2(a) - Building Tcl itself
 
@@ -256,14 +256,14 @@
 TCLINSTALL = 0 # Tk always builds against Tcl source, not an installed Tcl
 !if "$(TCLDIR)" == ""
 !if [echo TCLDIR = \> nmakehlp.out] \
-   || [nmakehlp -L generic\tcl.h >> nmakehlp.out]
+   || [nmakehlp -L $(_TCL_H) >> nmakehlp.out]
 !error *** Could not locate Tcl source directory.
 !endif
 !include nmakehlp.out
 !endif # TCLDIR == ""
 
 _TCLDIR	= $(TCLDIR:/=\)
-_TCL_H  = $(_TCLDIR)\generic\tcl.h
+_TCL_H  = $(_TCLDIR)\include\tcl.h
 !if !exist("$(_TCL_H)")
 !error Could not locate tcl.h. Please set the TCLDIR macro to point to the Tcl *source* directory.
 !endif
@@ -284,9 +284,9 @@
 !if exist("$(_TCLDIR)\include\tcl.h") # Case 2(c) with TCLDIR defined
 TCLINSTALL	= 1
 _TCL_H          = $(_TCLDIR)\include\tcl.h
-!elseif exist("$(_TCLDIR)\generic\tcl.h") # Case 2(d) with TCLDIR defined
+!elseif exist("$(_TCLDIR)\include\tcl.h") # Case 2(d) with TCLDIR defined
 TCLINSTALL	= 0
-_TCL_H          = $(_TCLDIR)\generic\tcl.h
+_TCL_H          = $(_TCLDIR)\include\tcl.h
 !endif
 
 !else  #  # Case 2(c) for extensions with TCLDIR undefined
@@ -306,13 +306,13 @@
 !else # exist(...) && !$(NEED_TCL_SOURCE)
 
 !if [echo _TCLDIR = \> nmakehlp.out] \
-   || [nmakehlp -L generic\tcl.h >> nmakehlp.out]
+   || [nmakehlp -L $(_TCL_H) >> nmakehlp.out]
 !error *** Could not locate Tcl source directory.
 !endif
 !include nmakehlp.out
 TCLINSTALL      = 0
 TCLDIR         = $(_TCLDIR)
-_TCL_H          = $(_TCLDIR)\generic\tcl.h
+_TCL_H          = $(_TCLDIR)\include\tcl.h
 
 !endif # exist(...) && !$(NEED_TCL_SOURCE)
 
@@ -627,7 +627,7 @@
 # generated libraries only usable by the specific VC++ version that
 # created it. Requires /LTCG linker option
 !if [nmakehlp -c -GL]
-OPTIMIZATIONS  = $(OPTIMIZATIONS) -GL
+
 CC_GL_OPT_ENABLED = 1
 !else
 # In newer compilers -GL and -YX are incompatible.
@@ -691,6 +691,13 @@
 # VERSION - set as (for example 25)
 #--------------------------------------------------------------
 
+!if !exists($(_TCL_H))
+!error Error: Cannot parse $(_TCL_H): file not found
+!endif
+!if !exists($(_TK_H))
+!error Error: Cannot parse $(_TK_H): file not found
+!endif
+
 !if [echo REM = This file is generated from rules.vc > versions.vc]
 !endif
 !if [echo TCL_MAJOR_VERSION = \>> versions.vc] \
@@ -1145,7 +1152,7 @@
 !if $(DOING_TCL)
 TCLSHNAME       = $(PROJECT)sh$(VERSION)$(SUFX).exe
 TCLSH		= $(OUT_DIR)\$(TCLSHNAME)
-TCLIMPLIB	= $(OUT_DIR)\$(PROJECT)$(VERSION)$(SUFX).lib
+
 TCLLIBNAME	= $(PROJECT)$(VERSION)$(SUFX).$(EXT)
 TCLLIB		= $(OUT_DIR)\$(TCLLIBNAME)
 TCLSCRIPTZIP    = $(OUT_DIR)\$(TCL_ZIP_FILE)
@@ -1179,7 +1186,7 @@
 # When building extensions, may be linking against Tcl that does not add
 # "t" suffix (e.g. 8.5 or 8.7). If lib not found check for that possibility.
 !if !exist("$(TCLIMPLIB)")
-TCLIMPLIB	= $(_TCLDIR)\lib\tcl$(TCL_VERSION)t$(SUFX:t=).lib
+
 !endif
 TCL_LIBRARY	= $(_TCLDIR)\lib
 TCLREGLIB	= $(_TCLDIR)\lib\tclreg13$(SUFX:t=).lib
@@ -1203,14 +1210,14 @@
 # When building extensions, may be linking against Tcl that does not add
 # "t" suffix (e.g. 8.5 or 8.7). If lib not found check for that possibility.
 !if !exist("$(TCLIMPLIB)")
-TCLIMPLIB	= $(_TCLDIR)\win\$(BUILDDIRTOP)\tcl$(TCL_VERSION)t$(SUFX:t=).lib
+
 !endif
 TCL_LIBRARY	= $(_TCLDIR)\library
 TCLREGLIB	= $(_TCLDIR)\win\$(BUILDDIRTOP)\tclreg13$(SUFX:t=).lib
 TCLDDELIB	= $(_TCLDIR)\win\$(BUILDDIRTOP)\tcldde14$(SUFX:t=).lib
 TCLSCRIPTZIP	= $(_TCLDIR)\win\$(BUILDDIRTOP)\$(TCL_ZIP_FILE)
 TCLTOOLSDIR	= $(_TCLDIR)\tools
-TCL_INCLUDES	= -I"$(_TCLDIR)\generic" -I"$(_TCLDIR)\win"
+TCL_INCLUDES	= -I"$(_TCLDIR)\include" -I"$(_TCLDIR)\win"
 
 !endif # TCLINSTALL
 
@@ -1525,7 +1532,6 @@
 
 !if $(DEBUG)
 # Turn warnings into errors
-cwarn = $(cwarn) -WX
 !endif
 
 INCLUDES = $(TCL_INCLUDES) $(TK_INCLUDES) $(PRJ_INCLUDES)
